glp/neighborlist.py

```diff
@@ -189,23 +212,210 @@ def candidates_fn(n):
     others = jnp.reshape(square, (-1,))
     return centers, others
 
+def cell_list_candidate_fn(bin_id, N, dim):
+    """Get the candidates for the cell list neighbor list. It is implemented following the jax-md implementation."""
+    idx = bin_id
+    bin_idx = [idx] * (dim**3)
+    # Get the neighboring bins from the current bin
+    for i, dindex in enumerate(neighboring_bins(dim)):
+        bin_idx[i] = shift_array(idx, dindex)
+
+    bin_idx = jnp.concatenate(bin_idx, axis=-2)
+    bin_idx = bin_idx[..., jnp.newaxis, :, :]
+    bin_idx = jnp.broadcast_to(bin_idx, idx.shape[:-1] + bin_idx.shape[-2:])
+    def copy_values_from_bin(value, bin_value, bin_id):
+        scatter_indices = jnp.reshape(bin_id, (-1,))
+        bin_value = jnp.reshape(bin_value, (-1,) + bin_value.shape[-2:])
+        return value.at[scatter_indices].set(bin_value)
+    neighbor_idx = jnp.zeros((N + 1,) + bin_idx.shape[-2:], jnp.int32)
+    neighbor_idx = copy_values_from_bin(neighbor_idx, bin_idx, idx)
+    # Reshape the neighbor_idx to get the centers and others
+    others = jnp.reshape(neighbor_idx[:-1, :, 0], (-1,))
+    centers = jnp.repeat(jnp.arange(0, N), others.shape[0] // N)
+    return centers, others
 
-def get_neighbors(positions, cell, square_distances, cutoff, padding_mask=None):
-    centers, others = candidates_fn(positions.shape[0])
-    sq_distances = square_distances(positions[centers], positions[others])
-
-    mask = sq_distances <= cutoff
-    mask = mask * (centers != others)  # remove self-interactions
+def get_neighbors(positions, square_distances, cutoff, padding_mask=None, cl=None):
+    N, dim = positions.shape
+    if cl is not None:
+        centers, others = cell_list_candidate_fn(cl.id, N, dim)
+    else:
+        centers, others = candidates_fn(N)
 
-    # mask out fake positions
+    
+    sq_distances = square_distances(positions[centers], positions[others])
+    mask = sq_distances <= (cutoff**2)
+    mask = mask * ((centers != others) & (others<N))  # remove self-interactions and neighbors repetitions
```

Comment:
In the initial allocation, the script assigns memory to each bin based on the sum of `bin_capacity` and a tolerance value (`extra_capacity`). This means that some bins may have more indices than there are atoms. The indices for the particles range from 0 to N-1, where N is the total number of atoms. The Nth index is used as a placeholder for an empty value. This placeholder must be masked to prevent duplicate  indices.
This is the line in which the placeholder are introduced:
` bin_id = N * jnp.ones((bin_count * bin_capacity, 1), dtype=jnp.int32)`